File: app/src/main/java/com/raredev/vcspace/ui/viewmodel/EditorViewModel.java
package com.raredev.vcspace.ui.viewmodel;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import com.raredev.vcspace.models.DocumentModel;
import java.util.ArrayList;
import java.util.List;

public class EditorViewModel extends ViewModel {

  private MutableLiveData<List<DocumentModel>> documents = new MutableLiveData<>(new ArrayList<>());
  private MutableLiveData<Integer> currentPosition = new MutableLiveData<>(-1);

  private MutableLiveData<Boolean> drawerState = new MutableLiveData<>(false);

  public LiveData<List<DocumentModel>> getDocumentsLiveData() {
    return documents;
  }

  public List<DocumentModel> getDocuments() {
    return getDocumentsLiveData().getValue();
  }

  public void addDocument(DocumentModel document) {
    List<DocumentModel> documentList = getDocuments();
    documentList.add(document);

    documents.setValue(documentList);
  }

  public DocumentModel getDocument(int index) {
    return getDocuments().get(index);
  }

  public int indexOf(String path) {
    for (int i = 0; i < getOpenedDocumentCount(); i++) {
      DocumentModel temp = getDocument(i);
      if (path.equals(temp.getPath())) {
        return i;
      }
    }
    return -1;
  }

  public int indexOf(DocumentModel doc) {
    for (int i = 0; i < getOpenedDocumentCount(); i++) {
      DocumentModel temp = getDocument(i);
      if (doc.getPath().equals(temp.getPath())) {
        return i;
      }
    }
    return -1;
  }

  public int getOpenedDocumentCount() {
    return getDocuments().size();
  }

  public void observeDocuments(LifecycleOwner owner, Observer<List<DocumentModel>> obs) {
    getDocumentsLiveData().observe(owner, obs);
  }

  public void removeDocument(int index) {
    List<DocumentModel> documentList = getDocuments();
    documentList.remove(index);
    documents.setValue(documentList);
  }

  public void clearDocuments() {
    List<DocumentModel> documents = getDocuments();
    documents.clear();
    this.documents.setValue(documents);
  }

  public LiveData<Boolean> getDrawerStateLiveData() {
    return drawerState;
  }

  public void setCurrentPosition(int position) {
    currentPosition.setValue(position);
  }

  public int getCurrentPosition() {
    return currentPosition.getValue();
  }

  public DocumentModel getCurrentDocument() {
    int currentPosition = getCurrentPosition();
    if (currentPosition == -1) {
      return null;
    }
    return getDocuments().get(currentPosition);
  }

  public void observeCurrentPosition(LifecycleOwner owner, Observer<Integer> obs) {
    currentPosition.observe(owner, obs);
  }

  public boolean getDrawerState() {
    return getDrawerStateLiveData().getValue();
  }

  public void setDrawerState(boolean state) {
    drawerState.setValue(state);
  }

  public void observeDrawerState(LifecycleOwner owner, Observer<Boolean> obs) {
    drawerState.observe(owner, obs);
  }
}

File: app/src/main/java/com/raredev/vcspace/ui/viewmodel/EditorViewModel.kt
package com.raredev.vcspace.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.Observer
import java.io.File
import java.util.ArrayList
import java.util.List

class EditorViewModel : ViewModel() {
  internal val _displayedFile = MutableLiveData(-1)

  private val editorFiles = MutableLiveData<MutableList<File>>(ArrayList())

  private val mCurrentFile = MutableLiveData<Pair<Int, File?>?>(null)

  var displayedFileIndex: Int
    get() = _displayedFile.value!!
    set(value) {
      _displayedFile.value = value
    }

  fun getDisplayedFile(): MutableLiveData<Int> {
    return _displayedFile
  }

  fun setCurrentFile(index: Int, file: File?) {
    mCurrentFile.value = index to file
    displayedFileIndex = index
  }

  fun addFile(file: File) {
    val files = editorFiles.value ?: mutableListOf()
    files.add(file)
    editorFiles.value = files
  }

  fun removeFile(index: Int) {
    val files = editorFiles.value ?: mutableListOf()
    files.removeAt(index)
    editorFiles.value = files
  }

  fun removeAllFiles() {
    editorFiles.value = ArrayList()
  }

  fun getOpenedFileCount(): Int {
    return editorFiles.value!!.size
  }

  fun getOpenedFiles(): MutableList<File> {
    return editorFiles.value!!
  }

  fun observeFiles(lifecycleOwner: LifecycleOwner?, observer: Observer<MutableList<File>?>?) {
    editorFiles.observe(lifecycleOwner!!, observer!!)
  }

  fun getCurrentFileIndex(): Int {
    return mCurrentFile.value?.first ?: -1
  }

  fun getCurrentFile(): File? {
    return mCurrentFile.value?.second
  }
}